popovers: Change style of emoji and giphy search bar.

nimishmedatwal opened the pull request:
<!-- Describe your pull request here.-->
This pull request focuses on enhancing the appearance of emoji and Giphy popovers by bringing their styling in line with the established `dropdown-widget style`, both in dark and light modes. This improvement was identified by @amanagr in [issue #26929](https://github.com/zulip/zulip/issues/26929).

Additionally, this PR introduces a new clear button in the emoji popover and  adds new functionality in clear button  by which it will only be visible when there is text present in the search box.

Also the glitch in giphy popover where you have to press tab 2 times to focus different gifs is fixed.


Fixes: #26929<!-- Issue link, or clear description.-->


<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.
Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->




**Screenshots and screen captures:**

![Animation3](https://github.com/zulip/zulip/assets/91622060/ffc03f2d-7005-4b4c-8c44-2f5813dc9ef2)

|  Without Text | With Text |
| --------------- | --------- |
| ![image](https://github.com/zulip/zulip/assets/91622060/fb527288-b6fa-427c-a510-e22c24e069fa) | ![image](https://github.com/zulip/zulip/assets/91622060/6fee8225-6e8f-4a3b-ad9a-35e9c00ba528) |
| ![image](https://github.com/zulip/zulip/assets/91622060/8743f711-7134-4f4d-b9d9-89c11545cb4b) | ![image](https://github.com/zulip/zulip/assets/91622060/31e7e62c-4ed1-481e-ad50-f98dd9d552f0) |





|        Before         |        After        |
| ------------------ | ------------- |
 Light Mode 
| ![image](https://github.com/zulip/zulip/assets/91622060/fd785c2b-b455-4935-b856-ccd0f402abb9) | ![image](https://github.com/zulip/zulip/assets/91622060/12d06c1c-4a98-4e85-b989-c05d6fd24d0e) |
| ![image](https://github.com/zulip/zulip/assets/91622060/6992265c-1506-4572-81c2-e04ca6734bc1) | ![image](https://github.com/zulip/zulip/assets/91622060/20161961-0fcc-40ea-8bf8-433b34f1fac9)|
Dark Mode
| ![image](https://github.com/zulip/zulip/assets/91622060/1719f84a-2f84-4908-ac70-8288418134fc) | ![image](https://github.com/zulip/zulip/assets/91622060/d0a2dcd5-bd04-49d4-88ba-4539ca8bfb3b)  |
| ![image](https://github.com/zulip/zulip/assets/91622060/b2a0c103-e006-494e-a8fc-adb2fc5ec998) | ![image](https://github.com/zulip/zulip/assets/91622060/a1ee1e28-4098-4d0c-8cee-d334621300b9)


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
Thanks! Can you add screenshots with text for both popovers, in addition to the ones you have without?

alya commented:
We should not have an `x` in the GIPHY search box when there is no text there.

nimishmedatwal commented:
> Thanks! Can you add screenshots with text for both popovers, in addition to the ones you have without?

Sure, Here they are:
![image](https://github.com/zulip/zulip/assets/91622060/7df67eb5-a0be-435e-8066-e8eedf60853c)
![image](https://github.com/zulip/zulip/assets/91622060/b103d272-9736-482e-bd8a-650eae49ac6a)

nimishmedatwal commented:
> We should not have an `x` in the GIPHY search box when there is no text there.
We can remove the `x` altogether as it is not there in emoji nor dropdown-widget popover.

nimishmedatwal commented:
> I think there should be an `x` in both of them when there's text there, and not when there isn't. That's our standard approach across the Zulip app.

Okay, will fix it. Also I noticed there are several other search bars where `x` is there whether there's any text or not like search usernames, Search all messages, filter streams etc. So I am thinking either I can create another pull request fixing all these or I can make changes in this pr only, what do you think will be better?

karlstolley commented:
@nimishmedatwal it would be really helpful to have before and after screenshots here, too, so it's more obvious what's being changed. Table syntax is really great for showing side-by-side before and after versions.

nimishmedatwal commented:
> I think there should be an `x` in both of them when there's text there, and not when there isn't. That's our standard approach across the Zulip app.

Actually I was thinking to add that functionality in another pull request, as I noticed that many of the search bars are having x even when there's no text, like search messages, search usernames and filter streams.
![image](https://github.com/zulip/zulip/assets/91622060/24bc97ff-15d0-41d5-a129-6d0626f52497)
![image](https://github.com/zulip/zulip/assets/91622060/91766402-c84c-4b64-b15d-70be5924137f)
![image](https://github.com/zulip/zulip/assets/91622060/275f18c7-b18c-4aab-8ba2-5eed8b2aca81)

alya commented:
Those search bars are closed by `x`, which is not a use for `x` that we have here.

nimishmedatwal commented:
> We should not have an `x` in the GIPHY search box when there is no text there.

@alya I've added that functionality, can you please give it another review?

alya commented:
Please update the screenshots in the PR description, showing boxes with and without text. Videos are quite difficult for me to review.

nimishmedatwal commented:
> Please update the screenshots in the PR description, showing boxes with and without text. Videos are quite difficult for me to review.

Updated !

alya commented:
This PR breaks `tab` keyboard navigation from the search box input to the `x`.

alya commented:
Works for me other than that.

nimishmedatwal commented:
> This PR breaks `tab` keyboard navigation from the search box input to the `x`.

@alya I've fixed `tab` keyboard navigation in the giphy search box input.
Screenshots :
![image](https://github.com/zulip/zulip/assets/91622060/6636e8d8-690e-4126-901c-85f41a713803)

![Animation](https://github.com/zulip/zulip/assets/91622060/d4f08694-eb7a-48e9-8df2-16fa81f65799)

I am hesitant to add this tab navigation in emoji search box input as in `emoji-picker.js` keydown event on tab is used to select the selected emoji by this code. Although if needed I'll remove this and change it to clear-button. So let me know what you think.

``` 
if (event_name === "tab") {
    $selected_emoji.trigger("focus");
    update_emoji_showcase($selected_emoji);
    return true;
}
```

alya commented:
What if we make the first tab go to the `x`, and tabbing when `x` is selected would go to the first emoji? That would probably feel pretty reasonable.

nimishmedatwal commented:
> What if we make the first tab go to the `x`, and tabbing when `x` is selected would go to the first emoji? That would probably feel pretty reasonable.
> 
> We want the same behavior in the GIPHY popover as well. The current behavior there seems weird -- I'm needing to press `Tab` more than once to get to each next location...

@alya actually that is because tabindex of simplebar-content-wrapper is set to 0 so it's focusing to that wrapper when pressing tab.
![image](https://github.com/zulip/zulip/assets/91622060/303ee19d-429c-4c92-98e1-6dac256b9868)
![image](https://github.com/zulip/zulip/assets/91622060/b21d08e8-f2b7-4d8e-b3b7-2cb2ca65c0d8)

nimishmedatwal commented:
@alya I've corrected that and now the emoji and giphy popovers tab focus are working perfectly fine.
![Animation](https://github.com/zulip/zulip/assets/91622060/4e4118db-42ec-44dc-86b2-4c9196800cd9)

alya commented:
@amanagr I'd like to pass this on to you for further manual testing and code review. Not a release priority.

amanagr commented on web/src/emoji_picker.js:
```diff
@@ -320,6 +320,32 @@ function process_enter_while_filtering(e) {
         }
     }
+
+    const $searchInput = $("#emoji-search-query");
+    const $clearButton = $("#emoji_search_clear");
+    $(document).on("input", "#emoji-search-query", () => {
+        if ($searchInput.val() === "") {
+            $clearButton.css("display", "none");
+        } else {
+            $clearButton.css("display", "block");
+        }
+    });
+
+    $(document).on("click", "#emoji_search_clear", async (e) => {
+        e.stopPropagation();
+        $("#emoji-search-query").val("");
+        show_emoji_catalog();
+        $clearButton.css("display", "none");
+    });
+    $(document).on("keydown", "#emoji_search_clear", async (e) => {
+        e.stopPropagation();
+        const id = e.key || e.which || 0;
+        if (id === 13) {
+            $("#emoji-search-query").val("");
+            show_emoji_catalog();
+            $clearButton.css("display", "none");
+        }
+    });
```
These don't belong inside `process_enter_while_filtering`. Let's place them with other event handlers.

amanagr replied:
Also, we prefer snake_case to camelCase for variable names.

amanagr commented on web/src/emoji_picker.js:
```diff
@@ -346,6 +372,7 @@ function reset_emoji_showcase() {
 function update_emoji_showcase($focused_emoji) {
     // Don't use jQuery's data() function here. It has the side-effect
     // of converting emoji names like :100:, :1234: etc to number.
+    $(".emoji-popover .simplebar-content-wrapper").attr("tabindex", "-1");
```
Can you add a comment here for the purpose of this? I assume this is to avoid focus on the simplebar wrapper.

amanagr commented on web/src/emoji_picker.js:
```diff
@@ -444,6 +471,10 @@ function change_focus_to_filter() {
     }
     reset_emoji_showcase();
 }
+function change_focus_to_clear() {
```
Let's rename this to `change_focus_to_clear_button`.

amanagr commented on web/src/giphy.js:
```diff
@@ -206,12 +206,23 @@ function toggle_giphy_popover(target) {
 
+                const $searchInput = $("#giphy-search-query");
+                const $clearButton = $("#giphy_search_clear");
+                $(document).on("input", "#giphy-search-query", () => {
+                    if ($searchInput.val().trim() === "") {
```
Let's keep this event handler scoped to `$popper` like other event handlers. Also, similar comment about using snake_case here.

amanagr commented on web/src/giphy.js:
```diff
@@ -206,12 +206,23 @@ function toggle_giphy_popover(target) {
+                    $clearButton.css("display", "none");
                 });
-
+                $("#giphy_grid_in_popover .simplebar-content-wrapper").attr("tabindex", "-1");
```
Needs a comment.

amanagr commented on web/styles/dark_theme.css:
```diff
@@ -516,6 +516,31 @@
+            position: sticky;
+            padding: 2px;
+
+            & input {
+                background: hsl(225deg 6% 10%);
+                color: hsl(0deg 0% 75%);
+            }
+
+            & input:focus {
```
Let's directly use the id/class of input element instaed of using the tag name.

amanagr reviewed (commented):
@nimishmedatwal thanks for working on this. Let's some comments. Looks great otherwise!

amanagr commented on web/src/giphy.js:
```diff
@@ -115,7 +115,7 @@ async function renderGIPHYGrid(targetEl) {
                     // user to navigate the GIFs using tab.
                     // TODO: Remove this after https://github.com/Giphy/giphy-js/issues/174
                     // is closed.
-                    e.target.tabIndex = 0;
+                    e.target.tabIndex = -1;
                 },
```
Why do need this? Also, looks like https://github.com/Giphy/giphy-js/issues/174 is closed, so we can just upgrade our version and remove this.

nimishmedatwal replied:
I am not sure how to upgrade the version of giphy-js. Can you guide me a bit on how to do it?

amanagr replied:
You need to update it in `package.json`.

nimishmedatwal replied:
@amanagr Updated!

nimishmedatwal commented:
@amanagr made the requested changes, kindly give it another review.

amanagr commented on web/src/giphy.js:
```diff
@@ -109,14 +109,6 @@ async function renderGIPHYGrid(targetEl) {
                     );
                     hide_giphy_popover();
                 },
-                onGifVisible(_gif, e) {
-                    // Set tabindex for all the GIFs that
-                    // are visible to the user. This allows
-                    // user to navigate the GIFs using tab.
-                    // TODO: Remove this after https://github.com/Giphy/giphy-js/issues/174
-                    // is closed.
-                    e.target.tabIndex = 0;
-                },
```
Let's extract this change and the giphy upgrade into an initial prep commit.

amanagr replied:
You also need to increase `PROVISION_VERSION` in `version.py`

amanagr commented on web/styles/dark_theme.css:
```diff
@@ -474,6 +474,31 @@
         color: hsl(236deg 33% 90%);
     }
 
+    #giphy_grid_in_popover {
+        .search-box {
+            display: flex;
+            position: sticky;
+            padding: 2px;
+
+            & #giphy-search-query {
+                background: hsl(225deg 6% 10%);
+                color: hsl(0deg 0% 75%);
+            }
+
+            & #giphy-search-query:focus {
+                background-color: hsl(225deg 6% 7%);
+                border: 1px solid hsl(0deg 0% 100% / 50%);
+                box-shadow: 0 0 5px hsl(0deg 0% 100% / 40%);
+            }
+        }
+    }
+
+    .emoji-popover-filter:focus {
+        background-color: hsl(225deg 6% 7%) !important;
+        border: 1px solid hsl(0deg 0% 100% / 50%) !important;
+        box-shadow: 0 0 5px hsl(0deg 0% 100% / 40%) !important;
+    }
```
We are migrating styles away from `dark_theme.css`. Let's use variables for colors and use them `popovers.css`. Color variables are defined in `zulip.css`.

karlstolley replied:
Just a note that CSS variables are now defined in `app_variables.css`

nimishmedatwal commented:
@amanagr made the requested changes!

amanagr commented:
@nimishmedatwal sorry for the late followup, can you rebase this?

amanagr commented:
![image](https://github.com/zulip/zulip/assets/25124304/01c2b9d3-fc25-441b-94be-eaa15a7a7790)

The search icon is not aligned in the center in both these popovoers, can you please fix it too?

amanagr commented on web/src/emoji_picker.js:
```diff
@@ -286,7 +287,7 @@ function toggle_reaction(emoji_name, event) {
         return;
     }
 
-    const message_id = current_message_id;
+    const message_id = message_lists.current.selected_id();
     const message = message_store.get(message_id);
```
I don't think this is a correct change since we are tracking for which message id the popover is open for which can be different from the selected message.

nimishmedatwal replied:
Since these changes were made quite some time ago, I can't recall why exactly I did them. For now, I've reverted them back as they appeared to have no impact on the functionality of my pull request.

amanagr commented on web/src/emoji_picker.js:
```diff
@@ -815,10 +856,10 @@ function register_click_handlers() {
     });
 
-    $("body").on("click", "#set-user-status-modal #selected_emoji .status-emoji-wrapper", (e) => {
+    $("body").on("click", "#set-user-status-modal #selected_emoji", (e) => {
         e.preventDefault();
         e.stopPropagation();
-        toggle_emoji_popover(e.currentTarget, undefined, {placement: "bottom"});
+        toggle_emoji_popover(e.target, undefined, {placement: "bottom"});
         if (is_open()) {
```
What is the motivatoin for this chnage?

amanagr commented on web/src/emoji_picker.js:
```diff
@@ -830,12 +871,6 @@ function register_click_handlers() {
-    $("body").on(
-        "keydown",
-        "#set-user-status-modal #selected_emoji .status-emoji-wrapper",
-        ui_util.convert_enter_to_click,
-    );
 }
```
IS this no longer required?

nimishmedatwal commented:
@amanagr I am having my mid semester exams right now so if this issue is not high priority would it alright if I work on it after 22nd March ?

nimishmedatwal commented:
@amanagr I've fixed the search icon in emoji popover, also since these changes were made quite some time ago, I can't recall why exactly I did them. For now, I've reverted them back as they had no impact on the functionality of my pull request.

nimishmedatwal commented:
Why is `E: Package 'libmagic1' has no installation candidate` this error coming in Installing dependencies. I've not made any changes related to that.

amanagr commented on web/styles/app_variables.css:
```diff
@@ -201,6 +201,13 @@
     --color-background-tab-picker-tab-option-active: hsl(0deg 0% 100% / 35%);
     --color-background-popover: hsl(0deg 0% 100%);
     --color-background-alert-word: hsl(18deg 100% 84%);
+    --color-background-giphy-search: hsl(0deg 0% 100%);
+    --color-background-giphy-search-focus: hsl(0deg 0% 100%);
+    --color-border-giphy-search-focus: hsl(229.09deg 21.57% 10% / 80%);
+    --color-box-shadow-giphy-search-focus: hsl(228deg 9.8% 20% / 30%);
+    --color-background-emoji-filter-focus: hsl(0deg 0% 100%);
+    --color-border-emoji-filter-focus: hsl(229.09deg 21.57% 10% / 80%);
+    --color-box-shadow-emoji-filter-focus: hsl(228deg 9.8% 20% / 30%);
```
Since all the colors are same, we can use a common variable here.

amanagr commented:
@nimishmedatwal thanks for the update! Left a comment, otherwise LGTM.

nimishmedatwal commented:
@amanagr I've changed the code so that it uses a common variable wherever possible. Thank you for your thorough review.

amanagr commented on web/src/emoji_picker.js:
```diff
@@ -592,6 +602,33 @@ function register_popover_events($popover) {
+        const $search_input = $("#emoji-search-query");
+        const $clear_button = $("#emoji_search_clear");
+        $(document).on("input", "#emoji-search-query", () => {
+            if ($search_input.val() === "") {
+                $clear_button.css("display", "none");
+            } else {
+                $clear_button.css("display", "block");
+            }
+        });
+
+        $(document).on("click", "#emoji_search_clear", async (e) => {
+            e.stopPropagation();
+            $("#emoji-search-query").val("");
+            show_emoji_catalog();
+            $clear_button.css("display", "none");
+        });
+        $(document).on("keydown", "#emoji_search_clear", async (e) => {
+            e.stopPropagation();
+            const id = e.key || e.which || 0;
+            if (id === 13) {
+                $("#emoji-search-query").val("");
+                show_emoji_catalog();
+                $clear_button.css("display", "none");
+            }
```
Wouldn't just using `hide` and `show` jQuery methods have the same effect?

amanagr commented:
@nimishmedatwal left a minor comment, otherwise LGTM.

nimishmedatwal commented:
@amanagr now using `hide` and `show`  jQuery methods.

amanagr commented:
LGTM! Thanks!

sayamsamal commented:
Hey @nimishmedatwal, how's the progress on this? Please let me know if you're stuck anywhere and would need some help. Otherwise, we could have one more look at the code and move this forward to be merged.

nimishmedatwal commented:
@sayamsamal I've resolved the conflicts and this pr is ready to be reviewed again!

sayamsamal commented:
Thanks @nimishmedatwal. I'll review this very soon.

sayamsamal commented on package.json:
```diff
@@ -10,6 +10,7 @@
     "@formatjs/intl": "^2.0.0",
     "@giphy/js-components": "^5.0.5",
     "@giphy/js-fetch-api": "^5.3.0",
+    "@koa/bodyparser": "^5.0.0",
```
The `@koa/bodyparser` is removed in the previous commit but the same version of that package is added back here, without any bump in the `PROVISION_VERSION`. Why is that?

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -895,14 +928,18 @@ ul {
     /* 300px of GIPHY grid + 5px is the extra gutter space
      * between gif columns. */
-    width: 305px;
+    width: 307px;
```
The comment just above indicates that the width should be `305px`. Please fix the mismatch.

sayamsamal commented on web/styles/popovers.css:
```diff
@@ -895,14 +928,18 @@ ul {
     }
 
+    .giphy-gif:focus {
+        outline: -webkit-focus-ring-color auto 1px;
```
Ok, `-webkit-focus-ring-color` is a value I did not know about earlier. I gave [this article](https://remysharp.com/til/css/focus-ring-default-styles) a read, and I think we should also add the`Highlight` value to support Firefox.

sayamsamal replied:
We also prefer the `[ width || style || color ]` ordering for the shorthand values.

sayamsamal commented on web/src/emoji_picker.js:
```diff
@@ -601,6 +611,34 @@ function register_popover_events($popover) {
+
+        $(document).on("keydown", "#emoji_search_clear", async (e) => {
+            e.stopPropagation();
+            const id = e.key || e.which || 0;
+            if (id === 13) {
```
We use `keydown_util.is_enter_event` for this, it is a good utility function which also takes care of the IME (Input Method Editor).

sayamsamal commented on web/src/emoji_picker.js:
```diff
@@ -601,6 +611,34 @@ function register_popover_events($popover) {
+            e.stopPropagation();
+            const id = e.key || e.which || 0;
+            if (id === 13) {
+                $("#emoji-search-query").val("");
+                show_emoji_catalog();
+                $clear_button.hide();
```
To handle the `enter` key event, you can use `$(e.currentTarget).trigger("click")` instead. It helps prevent having to duplicate the same code all over again.

sayamsamal commented on web/src/emoji_picker.js:
```diff
@@ -601,6 +611,34 @@ function register_popover_events($popover) {
+    $(".emoji-popover-filter").on("keydown", () => {
+        const $search_input = $("#emoji-search-query");
+        const $clear_button = $("#emoji_search_clear");
+        $(document).on("input", "#emoji-search-query", () => {
```
Why are these event listeners nested? If I remember right, event listeners do apply to hidden elements since they do in fact exist in the DOM. If they didn't we could use event delegation to make it work, instead of these nested event listeners attached on the `document`.

sayamsamal commented on web/src/emoji_picker.js:
```diff
@@ -601,6 +611,34 @@ function register_popover_events($popover) {
+            }
+        });
+
+        $(document).on("click", "#emoji_search_clear", async (e) => {
```
Why are we using `async` over here? I don't see any I/O-bound or network-bound requests.

sayamsamal commented:
Okay, I have left you an initial review to keep you unblocked. It would also be nice to see https://github.com/Giphy/giphy-js/issues/174 mentioned in the commit message since that version bump actually fixed an issue for us, and also since as a bonus that issue was started by one of our maintainers.

I still haven't performed a manual review BTW. (mostly a reminder to myself)

amanagr commented:
Closing in favour of #30624
